core/shared/src/main/scala/spire/math/interval/Overlap.scala

```diff
@@ -0,0 +1,65 @@
+package spire.math.interval
+
+import spire.algebra.{Eq, Order}
+import spire.math.Interval
+import spire.math.interval.Overlap.{Disjoint, Equal, Subset}
+
+/**
+  * An ADT that represents overlapping result for any two intervals.
+  */
+sealed abstract class Overlap[A: Order] extends Serializable {
+  def isDisjoint: Boolean = this.isInstanceOf[Disjoint[_]]
+  def isSubset: Boolean = this match {
+    case Subset(_, _) | Equal() => true
+    case _ => false
+  }
+  def isEqual: Boolean = this.isInstanceOf[Equal[_]]
+}
+
+object Overlap {
```

Comment:
I should've thought of this in the last review (sorry!), but it feels like the `Interval#overlap` method implemented should live here instead as `apply[A: Order](a: Interval[A], b: Interval[A]): Overlap[A] = { ... }`. We'd still keep the `overlap` method on `Interval`, but it would just be implemented using `Overlap[A](this, that)` (eg `def overlap(that): Overlap[A] = Overlap(this, that)`).

Do you think that makes sense? Mainly, if we have this nice `Overlap` ADT, I feel like we should have some sort of public constructor on it and just a convenience method on `Interval` instead.